### PR TITLE
sign in

### PR DESCRIPTION
**Pull request by jonnyalexbh**
## Summary
- Sign in

## Trello Card
https://trello.com/c/U354g6S4/20-sign-in

## Screenshot
Captura de pantalla 2019-08-20 a la(s) 9 43 53 p  m

### CODE REVIEW

File: app/graphql/inputs.js [hunk 1]
```diff
@@ -5,8 +5,8 @@ module.exports = gql`
     firstName: String!
     lastName: String!
     username: String!
-    email: String! @constraint(format: "email", minLength: 12, pattern: ".*@wolox.(ar|co|cl|com.ar)$")
-    password: String! @constraint(pattern: "^[a-zA-Z0-9_]*$", minLength: 8)
```
Comment (emanuelcasco) [inline app/graphql/inputs.js]: Why do you remove this validations?

File: app/graphql/types.js [hunk 2]
```diff
@@ -26,7 +27,7 @@ module.exports = gql`
   }
   type AccessToken {
     accessToken: String!
-    refreshToken: String!
-    expiresIn: Int!
+    refreshToken: String! @deprecated(reason: "it is not used")
+    expiresIn: Int! @deprecated(reason: "it is not used")
```
Comment (emanuelcasco) [inline app/graphql/types.js]: This fields should't be deprecated

File: app/helpers.js [hunk 3]
```diff
@@ -35,3 +40,12 @@ exports.changeUserIdByArtist = album => {
...
 };
+
+exports.generateToken = user => {
+  const tokenPayload = {
+    lastName: user.lastName,
+    firstName: user.firstName,
+    email: user.email
+  };
+  return jwt.encode(tokenPayload, secret_key);
+};
```
Comment (emanuelcasco) [inline app/helpers.js]: Move jwt encryption to a service, or create a helper specifically for it

**Review by emanuelcasco — CHANGES_REQUESTED** [review]

**Comment by emanuelcasco** [comment]
Don't forget writing test for new features!

File: app/graphql/inputs.js (hunk 1 above)
Comment (jonnyalexbh) [reply app/graphql/inputs.js]: because it presents problems in the scheme

Comment (jonnyalexbh) [reply app/graphql/inputs.js]: I added it again it works but it presents errors in the playground

File: app/helpers.js (hunk 3 above)
Comment (jonnyalexbh) [reply app/helpers.js]: requested changes ok

File: app/graphql/users/resolvers.js [hunk 4]
```diff
@@ -16,3 +17,24 @@ exports.createUser = async user => {
...
+
+exports.getUsers = () =>
+  User.getAll()
+    .then(foundUsers =>
+      foundUsers.map(foundUser => ({
+        ...foundUser.dataValues,
+        name: `${foundUser.firstName} ${foundUser.lastName}`
+      }))
+    )
+    .then(users => users);
```
Comment (emanuelcasco) [inline app/graphql/users/resolvers.js]: This is not necessary

File: app/graphql/users/resolvers.js [hunk 5]
```diff
@@ -16,3 +17,24 @@ exports.createUser = async user => {
       : errors.defaultError();
   }
 };
+
+exports.getUser = user =>
+  User.getOne(user).then(foundUser => {
+    foundUser.name = `${foundUser.firstName} ${foundUser.lastName}`;
```
Comment (emanuelcasco) [inline app/graphql/users/resolvers.js]: abstract this to a helper

File: app/graphql/users/resolvers.js [hunk 6]
```diff
@@ -16,3 +17,24 @@ exports.createUser = async user => {
...
+    return foundUser;
+  });
+
+exports.getUsers = () =>
+  User.getAll()
+    .then(foundUsers =>
+      foundUsers.map(foundUser => ({
+        ...foundUser.dataValues,
+        name: `${foundUser.firstName} ${foundUser.lastName}`
```
Comment (emanuelcasco) [inline app/graphql/users/resolvers.js]: Use the helper implemented above

File: app/services/jwt-encryption.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+const jwt = require('jwt-simple'),
+  config = require('../../config'),
+  { secret_key } = config.common.jwt;
+
+exports.generateToken = user => {
+  const tokenPayload = {
+    lastName: user.lastName,
+    firstName: user.firstName,
+    email: user.email
+  };
```
Comment (emanuelcasco) [inline app/services/jwt-encryption.js:9]: Why are you saving first and last name? Also, take in consideration that this values will be deprecated

File: app/services/user.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+  error = require('../errors'),
+  { generateToken } = require('./jwt-encryption');
+
+exports.checkLogin = credentials => {
+  const { username, password } = credentials;
+  return User.getByUsername(username).then(data => {
+    if (data && checkPassword(password, data.password)) {
+      return generateToken(data);
+    }
+    logger.error(error);
```
Comment (emanuelcasco) [inline app/services/user.js]: what are you logging here?

Comment (emanuelcasco) [reply app/services/user.js]: What happens if any error is thrown while executing this function (`checkLogin `)?

**Review by emanuelcasco — CHANGES_REQUESTED** [review]

File: app/graphql/users/resolvers.js (hunk 4 above)
Comment (jonnyalexbh) [reply app/graphql/users/resolvers.js]: it is understood

File: app/graphql/users/resolvers.js (hunk 5 above)
Comment (jonnyalexbh) [reply app/graphql/users/resolvers.js]: it is understood

File: app/graphql/users/resolvers.js [hunk 9]
```diff
@@ -16,3 +17,22 @@ exports.createUser = async user => {
       : errors.defaultError();
   }
 };
+
+exports.getUser = user =>
+  User.getOne(user).then(foundUser => {
+    foundUser.name = helpers.fullName(foundUser);
+    return foundUser;
+  });
```
Comment (emanuelcasco) [inline app/graphql/users/resolvers.js]: You could do:

```javascript
// helper
const addFullName = user => { ...user, name: `....`  };

// resolver
exports.getUser = user => User.getOne(user).then(addFullName);

exports.getUsers = () => User.getAll().then(users => users.map(addFullName));

```

**Review by emanuelcasco — CHANGES_REQUESTED** [review]

File: app/graphql/users/resolvers.js [hunk 10]
```diff
@@ -16,3 +17,8 @@ exports.createUser = async user => {
       : errors.defaultError();
   }
 };
+
+exports.signIn = async user => {
+  const token = await checkLogin(user);
+  return { accessToken: token };
```
Comment (emanuelcasco) [inline app/graphql/users/resolvers.js]: Directly name it `accessToken`, and return `{ accessToken }`

File: app/services/user.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+exports.checkLogin = credentials => {
+  const { username, password } = credentials;
+  return User.getByUsername(username)
+    .then(user => {
+      if (!user) {
+        throw errors.signInError('The username provided is incorrect');
+      }
+
+      if (!checkPassword(password, user.password)) {
+        throw errors.signInError('The password provided is incorrect');
```
Comment (emanuelcasco) [inline app/services/user.js:18]: Add log

File: app/services/user.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+  logger = require('../logger'),
+  errors = require('../errors'),
+  { generateToken } = require('./jwt-encryption');
+
+exports.checkLogin = credentials => {
+  const { username, password } = credentials;
+  return User.getByUsername(username)
+    .then(user => {
+      if (!user) {
+        throw errors.signInError('The username provided is incorrect');
```
Comment (emanuelcasco) [inline app/services/user.js:13]: Log

File: app/services/user.js [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+      }
+
+      if (!checkPassword(password, user.password)) {
+        throw errors.signInError('The password provided is incorrect');
+      }
+
+      return generateToken(user);
```
Comment (emanuelcasco) [inline app/services/user.js:21]: Log successfully authenticated

**Review by emanuelcasco — CHANGES_REQUESTED** [review]

File: app/services/jwt-encryption.js (hunk 7 above)
Comment (emanuelcasco) [reply app/services/jwt-encryption.js:9]: I still not understand why do  you need first and last name here. I would remove them, you just need `user` in order to search for the user when authenticating

**Review by emanuelcasco — CHANGES_REQUESTED** [review]

**Review by emanuelcasco — APPROVED** [review]

**Pull request merged**